### PR TITLE
v4: Convert %(perl: ) to new xlat API

### PR DESCRIPTION
**Pull request by ndptech**

### CODE REVIEW

File: src/modules/rlm_perl/rlm_perl.c [hunk 1]
```diff
@@ -424,10 +421,14 @@ static ssize_t perl_xlat(UNUSED TALLOC_CTX *ctx, char **out, size_t outlen,
 			(void)POPs;
 		} else if (count > 0) {
 			tmp = POPp;
-			strlcpy(*out, tmp, outlen);
-			ret = strlen(*out);
-
-			RDEBUG2("Len is %zu , out is %s freespace is %zu", ret, *out, outlen);
+			MEM(vb = fr_value_box_alloc_null(ctx));
+			if (fr_value_box_strdup(ctx, vb, NULL, tmp, true) < 0) {
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c:425]: should use bstrndup

File: src/modules/rlm_perl/rlm_perl.c [hunk 2]
```diff
@@ -351,6 +351,49 @@ static void xs_init(pTHX)
...
+			ssize_t	slen;
+
+			slen = fr_value_box_print_quoted(&FR_SBUFF_OUT(buffer, sizeof(buffer)), vb, T_BARE_WORD);
+			if (slen < 0) return -1;
+			sv = newSVpvn(buffer, (size_t)slen);
+		}
+			break;
+		}
+		if (!sv) return -1;
+		SvTAINT(sv);
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c:393]: Probably want to reflect the taint status of the individual boxes?

File: src/modules/rlm_perl/rlm_perl.c [hunk 3]
```diff
@@ -394,6 +394,123 @@ static int perl_vblist_to_av(AV *av, fr_value_box_list_t *head) {
...
+	HV		*hv;
+	I32		sv_len, i;
+	int		type;
+
+	type = SvTYPE(sv);
+
+	switch (type) {
+	case SVt_IV:
+	/*	Integer or Reference */
+		DEBUG3("Integer or Reference returned");
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c:427]: References are special, we should probably deref if it's an array and convert that.  Blessed references can also contain package names, so we could in theory support certain cpan libraries and their underlying types... maybe that's future work though.

File: src/modules/rlm_perl/rlm_perl.c [hunk 4]
```diff
@@ -394,6 +394,123 @@ static int perl_vblist_to_av(AV *av, fr_value_box_list_t *head) {
...
+		if (SvROK(sv)) {
+			if (perl_sv_to_vblist(ctx, list, request, SvRV(sv)) < 0) return -1;
+			break;
+		}
+		MEM(vb = fr_value_box_alloc(ctx, FR_TYPE_INT32, NULL, false));
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c:432]: DEBUG3("Integer")

File: src/modules/rlm_perl/rlm_perl.c [hunk 5]
```diff
@@ -394,6 +394,123 @@ static int perl_vblist_to_av(AV *av, fr_value_box_list_t *head) {
...
+	int		type;
+
+	type = SvTYPE(sv);
+
+	switch (type) {
+	case SVt_IV:
+	/*	Integer or Reference */
+		DEBUG3("Integer or Reference returned");
+		if (SvROK(sv)) {
+			if (perl_sv_to_vblist(ctx, list, request, SvRV(sv)) < 0) return -1;
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c:429]: DEBUG3("Reference returned")

File: src/modules/rlm_perl/rlm_perl.c [hunk 6]
```diff
@@ -394,6 +394,123 @@ static int perl_vblist_to_av(AV *av, fr_value_box_list_t *head) {
...
+	/*	Float */
+		DEBUG3("Float returned");
+		MEM(vb = fr_value_box_alloc(ctx, FR_TYPE_FLOAT64, NULL, false));
+		vb->vb_float64 = SvNV(sv);
+		break;
+
+	case SVt_PV:
+	/*	String */
+		DEBUG3("String returned");
+		tmp = SvPVbyte(sv, len);
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c:446]: `SvPVutf8`

File: src/tests/modules/perl/test.pl [hunk 7]
```diff
@@ -67,14 +67,14 @@ sub authenticate {
 	if ($RAD_REQUEST{'User-Name'} =~ /^baduser/i) {
 		# Reject user and tell him why
 		$RAD_REPLY{'Reply-Message'} = "Denied access by rlm_perl function";
-		return RLM_MODULE_REJECT;
+		return RLM_MODULE_OK;
```
Comment (arr2036) [inline src/tests/modules/perl/test.pl:70]: hm?

File: src/modules/rlm_perl/rlm_perl.c [hunk 8]
```diff
@@ -351,6 +351,166 @@ static void xs_init(pTHX)
...
+		switch (vb->type) {
+		case FR_TYPE_STRING:
+			sv = newSVpvn(vb->vb_strvalue, vb->length);
+			break;
+
+		case FR_TYPE_OCTETS:
+			sv = newSVpvn((char const *)vb->vb_octets, vb->vb_length);
+			break;
+
+		default:
```
Comment (arr2036) [inline src/modules/rlm_perl/rlm_perl.c:381]: Handling `FR_TYPE_GROUP` probably wouldn't hurt.

File: src/modules/rlm_perl/rlm_perl.c (hunk 4 above)
Comment (arr2036) [reply src/modules/rlm_perl/rlm_perl.c:432]: I think we can actually get the taint status from perl as well for all of these.

File: src/tests/modules/perl/test.pl (hunk 7 above)
Comment (ndptech) [reply src/tests/modules/perl/test.pl:70]: For the purpose of the test this was needed - returning RLM_MODULE_REJECT causes the test to stop immediately.  I'll change it to RLM_MODULE_NOTFOUND and add a comment

**Pull request merged**